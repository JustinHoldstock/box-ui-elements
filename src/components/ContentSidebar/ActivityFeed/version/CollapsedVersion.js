import PropTypes from 'prop-types';
import React from 'react';
import { FormattedMessage, injectIntl, intlShape } from 'react-intl';

import PlainButton from '../../../components/plain-button';
import IconInfoInverted from '../../../icons/general/IconInfoInverted';
import { UserPropType } from '../../../common/box-proptypes';

import messages from '../messages';

import './Version.scss';

function getMessageForAction(action, collaborators, versionStart, versionEnd) {
    // We only support collapsing for multiple upload versions
    if (action !== 'upload') {
        return null;
    }
    const collaboratorIDs = Object.keys(collaborators);
    const numberOfCollaborators = collaboratorIDs.length;

    if (numberOfCollaborators === 1) {
        const collaborator = collaborators[collaboratorIDs[0]];
        return (
            <FormattedMessage
                {...messages.versionUploadCollapsed}
                values={{
                    name: <strong>{collaborator.name}</strong>,
                    versions: (
                        <span className='box-ui-version-range'>
                            {versionStart} - {versionEnd}
                        </span>
                    )
                }}
            />
        );
    }

    return (
        <FormattedMessage
            {...messages.versionMultipleUsersUploaded}
            values={{
                numberOfCollaborators,
                versions: (
                    <span className='box-ui-version-range'>
                        {versionStart} - {versionEnd}
                    </span>
                )
            }}
        />
    );
}

const CollapsedVersion = ({ action, collaborators, intl, onInfo, versions, versionStart, versionEnd }) => (
    <div className='box-ui-collapsed-version'>
        <span className='box-ui-version-message'>
            {getMessageForAction(action, collaborators, versionStart, versionEnd)}
        </span>
        {onInfo ? (
            <span className='box-ui-version-actions'>
                <PlainButton
                    aria-label={intl.formatMessage(messages.getVersionInfo)}
                    className='box-ui-version-info'
                    onClick={() => {
                        onInfo({ versions });
                    }}
                    type='button'
                >
                    <IconInfoInverted height={16} width={16} />
                </PlainButton>
            </span>
        ) : null}
    </div>
);

CollapsedVersion.displayName = 'CollapsedVersion';

CollapsedVersion.propTypes = {
    action: PropTypes.oneOf(['upload']).isRequired,
    collaborators: PropTypes.objectOf(UserPropType).isRequired,
    intl: intlShape.isRequired,
    onInfo: PropTypes.func,
    versions: PropTypes.array,
    versionStart: PropTypes.number.isRequired,
    versionEnd: PropTypes.number.isRequired
};

export { CollapsedVersion as CollapsedVersionBase };
export default injectIntl(CollapsedVersion);
