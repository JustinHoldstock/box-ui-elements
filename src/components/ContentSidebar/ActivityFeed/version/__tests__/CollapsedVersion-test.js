import React from 'react';
import { shallow } from 'enzyme';

import { CollapsedVersionBase as CollapsedVersion } from '../CollapsedVersion';

const translationProps = {
    intl: { formatMessage: () => {} }
};

describe('features/activity-feed/version/CollapsedVersion', () => {
    const render = (item) => shallow(<CollapsedVersion {...translationProps} {...item} />);

    test('should correctly render for single collaborator', () => {
        const versionStart = 1;
        const versionEnd = 10;
        const item = {
            action: 'upload',
            collaborators: { 1: { name: 'Person one', id: 1 } },
            versionStart,
            versionEnd
        };

        const wrapper = render(item);
        const formattedMessage = wrapper.find('FormattedMessage');

        expect(wrapper.hasClass('box-ui-collapsed-version')).toBe(true);
        expect(formattedMessage.length).toBe(1);
        expect(formattedMessage.prop('values').name).toBeTruthy();
        expect(formattedMessage.prop('values').versions).toBeTruthy();

        const renderedVersionsMessage = shallow(formattedMessage.prop('values').versions);
        expect(renderedVersionsMessage.hasClass('box-ui-version-range')).toBe(true);
        expect(renderedVersionsMessage.text()).toEqual(`${versionStart} - ${versionEnd}`);
    });

    test('should correctly render for multiple collaborators', () => {
        const versionStart = 1;
        const versionEnd = 10;
        const item = {
            action: 'upload',
            collaborators: {
                1: { name: 'Person one', id: 1 },
                2: { name: 'Person two', id: 2 }
            },
            versionStart,
            versionEnd
        };

        const wrapper = render(item);
        const formattedMessage = wrapper.find('FormattedMessage');

        expect(wrapper.hasClass('box-ui-collapsed-version')).toBe(true);
        expect(formattedMessage.length).toBe(1);
        expect(formattedMessage.prop('values').numberOfCollaborators).toEqual(2);
        expect(formattedMessage.prop('values').versions).toBeTruthy();

        const renderedVersionsMessage = shallow(formattedMessage.prop('values').versions);
        expect(renderedVersionsMessage.hasClass('box-ui-version-range')).toBe(true);
        expect(renderedVersionsMessage.text()).toEqual(`${versionStart} - ${versionEnd}`);
    });

    test('should correctly render info icon if onInfo is passed', () => {
        const item = {
            action: 'upload',
            onInfo: () => {},
            collaborators: {
                1: { name: 'Person one', id: 1 },
                2: { name: 'Person two', id: 2 }
            },
            versionStart: 1,
            versionEnd: 10
        };

        const wrapper = render(item);

        expect(wrapper.find('IconInfoInverted').length).toBe(1);
    });

    test('should not render a message if action is not upload', () => {
        const item = {
            action: 'delete',
            collaborators: { 1: { name: 'Person one', id: 1 } },
            versionStart: 1,
            versionEnd: 10
        };

        const wrapper = render(item);
        const formattedMessage = wrapper.find('FormattedMessage');

        expect(wrapper.hasClass('box-ui-collapsed-version')).toBe(true);
        expect(formattedMessage.length).toBe(0);
    });
});
