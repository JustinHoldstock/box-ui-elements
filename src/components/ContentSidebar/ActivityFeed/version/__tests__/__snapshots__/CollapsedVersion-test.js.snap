// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/activity-feed/version/CollapsedVersion should correctly render for multiple collaborators 1`] = `
<div
  className="bcs-collapsed-version"
>
  <span
    className="bcs-version-message"
  >
    <FormattedMessage
      values={
        Object {
          "numberOfCollaborators": 2,
          "versions": <span
            className="bcs-version-range"
          >
            1
             - 
            10
          </span>,
        }
      }
    />
  </span>
</div>
`;

exports[`features/activity-feed/version/CollapsedVersion should correctly render for multiple collaborators 2`] = `
<span
  className="bcs-version-range"
>
  1
   - 
  10
</span>
`;

exports[`features/activity-feed/version/CollapsedVersion should correctly render for single collaborator 1`] = `
<div
  className="bcs-collapsed-version"
>
  <span
    className="bcs-version-message"
  >
    <FormattedMessage
      values={
        Object {
          "name": <strong>
            Person one
          </strong>,
          "versions": <span
            className="bcs-version-range"
          >
            1
             - 
            10
          </span>,
        }
      }
    />
  </span>
</div>
`;

exports[`features/activity-feed/version/CollapsedVersion should correctly render for single collaborator 2`] = `
<span
  className="bcs-version-range"
>
  1
   - 
  10
</span>
`;
